### PR TITLE
Modernize the ImageNet example

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

See comments in the code for changes.

This example is our version of https://github.com/pytorch/examples/blob/master/imagenet/main.py

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda

### CODE REVIEW

File: pl_examples/domain_templates/imagenet.py [hunk 1]
```diff
@@ -70,9 +65,8 @@ def __init__(
         lr: float = 0.1,
         momentum: float = 0.9,
         weight_decay: float = 1e-4,
-        batch_size: int = 4,
-        workers: int = 2,
-        **kwargs,
+        batch_size: int = 256,
+        workers: int = 4,
```
Comment (carmocca) [inline pl_examples/domain_templates/imagenet.py:69]: These now match the values in https://github.com/pytorch/examples/blob/master/imagenet/main.py

File: pl_examples/domain_templates/imagenet.py [hunk 2]
```diff
@@ -86,53 +80,56 @@ def __init__(
         self.workers = workers
         self.model = models.__dict__[self.arch](pretrained=self.pretrained)
 
+        self.train_acc1 = Accuracy(top_k=1)
+        self.train_acc5 = Accuracy(top_k=5)
+        self.eval_acc1 = Accuracy(top_k=1)
+        self.eval_acc5 = Accuracy(top_k=5)
```
Comment (carmocca) [inline pl_examples/domain_templates/imagenet.py:86]: Use torchmetrics

File: pl_examples/domain_templates/imagenet.py [hunk 3]
```diff
@@ -86,53 +80,56 @@ def __init__(
...
         return self.model(x)
 
     def training_step(self, batch, batch_idx):
         images, target = batch
         output = self(images)
         loss_train = F.cross_entropy(output, target)
-        acc1, acc5 = self.__accuracy(output, target, topk=(1, 5))
-        self.log("train_loss", loss_train, on_step=True, on_epoch=True, logger=True)
-        self.log("train_acc1", acc1, on_step=True, prog_bar=True, on_epoch=True, logger=True)
-        self.log("train_acc5", acc5, on_step=True, on_epoch=True, logger=True)
```
Comment (carmocca) [inline pl_examples/domain_templates/imagenet.py:99]: Let the `on_*` be inferred

File: pl_examples/domain_templates/imagenet.py [hunk 4]
```diff
@@ -160,91 +157,31 @@ def val_dataloader(self):
                 ),
             ),
             batch_size=self.batch_size,
-            shuffle=False,
             num_workers=self.workers,
+            pin_memory=True,
```
Comment (carmocca) [inline pl_examples/domain_templates/imagenet.py:161]: Pin memory

File: pl_examples/domain_templates/imagenet.py [hunk 5]
```diff
@@ -160,91 +157,31 @@ def val_dataloader(self):
...
-    )
-    parent_parser.add_argument("--seed", type=int, default=42, help="seed for initializing training.")
-    parser = ImageNetLightningModel.add_model_specific_args(parent_parser)
-    parser.set_defaults(profiler="simple", deterministic=True, max_epochs=90)
-    args = parser.parse_args()
-    main(args)
-
 
 if __name__ == "__main__":
-    cli_lightning_logo()
```
Comment (carmocca) [inline pl_examples/domain_templates/imagenet.py:249]: The logo is removed. Showing it is annoying because most people copy paste this file into their local system and this requires the entire repository as it is defined in the `pl_examples` directory

File: pl_examples/domain_templates/imagenet.py [hunk 6]
```diff
@@ -160,91 +157,31 @@ def val_dataloader(self):
...
-    run_cli()
+    LightningCLI(
+        ImageNetLightningModel,
+        trainer_defaults={
+            "max_epochs": 90,
+            "accelerator": "auto",
+            "devices": 1,
+            "logger": False,
```
Comment (carmocca) [inline pl_examples/domain_templates/imagenet.py:176]: Disable the logger as it is not part of the PyTorch example

File: pl_examples/domain_templates/imagenet.py [hunk 7]
```diff
@@ -160,91 +157,31 @@ def val_dataloader(self):
...
+            "benchmark": True,
```
Comment (carmocca) [inline pl_examples/domain_templates/imagenet.py:177]: Enable `benchmark`

File: pl_examples/domain_templates/imagenet.py [hunk 8]
```diff
@@ -160,91 +157,31 @@ def val_dataloader(self):
...
+            "callbacks": [
+                # the PyTorch example refreshes every 10 batches
+                TQDMProgressBar(refresh_rate=10),
+                # save when the validation top1 accuracy improves
+                ModelCheckpoint(monitor="val_acc1", mode="max"),
```
Comment (carmocca) [inline pl_examples/domain_templates/imagenet.py:182]: Mimic what's done in the PyTorch example

File: pl_examples/domain_templates/imagenet.py [hunk 9]
```diff
@@ -160,91 +157,31 @@ def val_dataloader(self):
...
-def run_cli():
-    parent_parser = ArgumentParser(add_help=False)
-    parent_parser = pl.Trainer.add_argparse_args(parent_parser)
-    parent_parser.add_argument("--data-path", metavar="DIR", type=str, help="path to dataset")
-    parent_parser.add_argument(
-        "-e", "--evaluate", dest="evaluate", action="store_true", help="evaluate model on validation set"
-    )
-    parent_parser.add_argument("--seed", type=int, default=42, help="seed for initializing training.")
-    parser = ImageNetLightningModel.add_model_specific_args(parent_parser)
-    parser.set_defaults(profiler="simple", deterministic=True, max_epochs=90)
```
Comment (carmocca) [inline pl_examples/domain_templates/imagenet.py:243]: We no longer profile and set deterministic by default as it slows down the code

File: pl_examples/domain_templates/imagenet.py [hunk 10]
```diff
@@ -86,53 +80,56 @@ def __init__(
...
     def configure_optimizers(self):
         optimizer = optim.SGD(self.parameters(), lr=self.lr, momentum=self.momentum, weight_decay=self.weight_decay)
         scheduler = lr_scheduler.LambdaLR(optimizer, lambda epoch: 0.1 ** (epoch // 30))
         return [optimizer], [scheduler]
 
+    def setup(self, stage=None):
+        if isinstance(self.trainer.strategy, ParallelStrategy):
+            # When using a single GPU per process and per `DistributedDataParallel`, we need to divide the batch size
+            # ourselves based on the total number of GPUs we have
+            self.batch_size = int(self.batch_size / max(1, self.trainer.devices))
```
Comment (rohitgr7) [inline pl_examples/domain_templates/imagenet.py:130]: maybe use num_processes?
I think if someone sets devices=-1, it stays as it is.. and here max(1, -1) won't split the batch_size as mentioned.
same for workers.

File: pl_examples/domain_templates/imagenet.py [hunk 11]
```diff
@@ -70,9 +65,8 @@ def __init__(
...
         momentum: float = 0.9,
         weight_decay: float = 1e-4,
-        batch_size: int = 4,
-        workers: int = 2,
-        **kwargs,
+        batch_size: int = 256,
+        workers: int = 4,
     ):
         super().__init__()
         self.save_hyperparameters()
```
Comment (rohitgr7) [inline pl_examples/domain_templates/imagenet.py:72]: should we remove some of the argument assignments below since most of them are set inside hparams already? plus down below batch_size is updated, so I think it won't get updated in hparams

File: pl_examples/domain_templates/imagenet.py [hunk 12]
```diff
@@ -86,53 +80,56 @@ def __init__(
...
+        self.train_acc5 = Accuracy(top_k=5)
+        self.eval_acc1 = Accuracy(top_k=1)
+        self.eval_acc5 = Accuracy(top_k=5)
+
     def forward(self, x):
         return self.model(x)
 
     def training_step(self, batch, batch_idx):
         images, target = batch
         output = self(images)
```
Comment (rohitgr7) [inline pl_examples/domain_templates/imagenet.py:93]:
```suggestion
        output = self.model(images)
```
just something we recommend to users.

File: pl_examples/domain_templates/imagenet.py [hunk 13]
```diff
@@ -86,53 +80,56 @@ def __init__(
...
+        # update metrics
+        self.train_acc1(output, target)
+        self.train_acc5(output, target)
+        self.log("train_acc1", self.train_acc1, prog_bar=True)
+        self.log("train_acc5", self.train_acc5, prog_bar=True)
         return loss_train
 
     def eval_step(self, batch, batch_idx, prefix: str):
         images, target = batch
         output = self(images)
```
Comment (rohitgr7) [inline pl_examples/domain_templates/imagenet.py:105]:
```suggestion
        output = self.model(images)
```

File: pl_examples/domain_templates/imagenet.py [hunk 14]
```diff
@@ -86,53 +80,56 @@ def __init__(
...
 
+    def setup(self, stage=None):
+        if isinstance(self.trainer.strategy, ParallelStrategy):
+            # When using a single GPU per process and per `DistributedDataParallel`, we need to divide the batch size
+            # ourselves based on the total number of GPUs we have
+            self.batch_size = int(self.batch_size / max(1, self.trainer.devices))
+            self.workers = int(self.workers / max(1, self.trainer.devices))
+
     def train_dataloader(self):
         normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
```
Comment (rohitgr7) [inline pl_examples/domain_templates/imagenet.py:134]: here it's creating dataset inside train_dataloader. although it works, but we recommend `setup` for these.
same for eval_dataloader

File: pl_examples/domain_templates/imagenet.py (hunk 11 above)
Comment (carmocca) [reply pl_examples/domain_templates/imagenet.py:72]: We can remove `save_hyperparemters` instead, as the CLI saves the config anyways. Having both is redundant

**Review by rohitgr7 — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Review by Borda — APPROVED** [review]
